libraries/tools/kotlin-symbol-processing-api/src/org/jetbrains/kotlin/ksp/utils.kt

```diff
@@ -99,33 +100,69 @@ fun KSDeclaration.isOpen() = !this.isLocal()
         || this.modifiers.contains(Modifier.ABSTRACT)
         || this.modifiers.contains(Modifier.OPEN)
         || (this.parentDeclaration as? KSClassDeclaration)?.classKind == ClassKind.INTERFACE
+        || (!this.modifiers.contains(Modifier.FINAL) && this.origin == Origin.JAVA)
         )
 
-fun KSDeclaration.isPublic() = !this.modifiers.contains(Modifier.PRIVATE)
-        && !this.modifiers.contains(Modifier.PROTECTED)
-        && !this.modifiers.contains(Modifier.INTERNAL)
+fun KSDeclaration.isPublic() = this.getVisibility() == Visibility.PUBLIC
+
+fun KSDeclaration.isProtected() = this.getVisibility() == Visibility.PROTECTED
 
 fun KSDeclaration.isInternal() = this.modifiers.contains(Modifier.INTERNAL)
 
 fun KSDeclaration.isPrivate() = this.modifiers.contains(Modifier.PRIVATE)
 
+fun KSDeclaration.isJavaPackagePrivate() = this.getVisibility() == Visibility.JAVA_PACKAGE
+
+fun KSDeclaration.closestClassDeclaration(): KSClassDeclaration? {
+    if (this is KSClassDeclaration) {
+        return this
+    } else {
+        return this.parentDeclaration?.closestClassDeclaration()
+    }
+}
+
 // TODO: cross module visibility is not handled
+// TODO: generics not handled, need API change to take type argument.
 fun KSDeclaration.isVisibleFrom(other: KSDeclaration): Boolean {
+    fun KSDeclaration.isSamePackage(other: KSDeclaration): Boolean = this.containingFile?.packageName == other.containingFile?.packageName
+
+    // lexical scope for local declaration.
+    fun KSDeclaration.parentDeclarationsForLocal(): List<KSDeclaration> {
+        val parents = mutableListOf<KSDeclaration>()
+
+        var parentDeclaration = this.parentDeclaration!!
+
+        while (parentDeclaration.isLocal()) {
+            parents.add(parentDeclaration)
+            parentDeclaration = parentDeclaration.parentDeclaration!!
+        }
+
+        parents.add(parentDeclaration)
+
+        return parents
+    }
+
+    fun KSDeclaration.isVisibleInPrivate(other: KSDeclaration) =
+        this.parentDeclaration == other.parentDeclaration
+                || this.parentDeclaration == other
+                || (
+                this.parentDeclaration == null
+                        && other.parentDeclaration == null
+                        && this.containingFile == other.containingFile
+                )
+
     return when {
         // locals are limited to lexical scope
-        this.isLocal() -> this.parentDeclaration == other
+        this.isLocal() -> this.parentDeclarationsForLocal().contains(other)
```

Comment:
nit: maybe add a todo for early break, rather than finding all containing declarations.